python/llm/test/inference_gpu/test_optimize_model.py

```diff
@@ -56,5 +80,115 @@ def test_optimize_model(Model, Tokenizer, model_path):
     assert any(diff) is False
 
 
+@pytest.mark.parametrize('Model, Tokenizer, model_path',[
+    (AutoModelForCausalLM, AutoTokenizer, os.environ.get('LLAMA_ORIGIN_PATH'))
+])
+
+def test_optimize_llama_model(Model, Tokenizer, model_path):
+    tokenizer = Tokenizer.from_pretrained(model_path, trust_remote_code=True)
+    input_ids = tokenizer.encode(prompt, return_tensors="pt").to(device)
+
+    model = Model.from_pretrained(model_path,
+                                load_in_4bit=True,
+                                optimize_model=False,
+                                trust_remote_code=True)
+    model = model.to(device)
+
+    for layer_name, layer_module in model.named_modules():
+        if layer_name == "model.layers.31.self_attn":
+            layer_module.register_forward_pre_hook(
+                lambda module, input: pre_hook(module, input))
+            layer_module.register_forward_hook(
+                lambda module, output, layer_name=layer_name: forward_hook(module, 
+                                                                           output, layer_name))
+    logits_base_model = (model(input_ids)).logits
+    layer_tensor = layer_outputs.pop()
+
+    del model
+
+    opt_model = Model.from_pretrained(model_path,
+                            load_in_4bit=True,
+                            optimize_model=True,
+                            trust_remote_code=True)
+    opt_model = opt_model.to(device)
+
+    for layer_name, layer_module in opt_model.named_modules():
+        if layer_name == "model.layers.31.self_attn":
+            layer_module.register_forward_pre_hook(
+                lambda module, input: load_pre_hook(module, input))
+            layer_module.register_forward_hook(
+                lambda module, output, layer_name=layer_name: forward_hook(module, 
+                                                                           output, layer_name))
+    logits_optimized_model = (opt_model(input_ids)).logits
+    opt_layer_tensor = layer_outputs.pop()
+
+    attn_output_diff = []
+    for i, (t1, t2) in enumerate(zip(layer_tensor, opt_layer_tensor)):
+        if t1 is not None and t2 is not None:
+            if not isinstance(t1, tuple) and not isinstance(t2, tuple):
+                attn_output_diff.append(t1 - t2)
+            else:
+                for i, (t3, t4) in enumerate(zip(t1, t2)):
+                    attn_output_diff.append(t3 - t4)
+
+    max_diff_tensor = [torch.max(item).item() for item in attn_output_diff]
+    assert all(max_diff <= lower_bound for max_diff in max_diff_tensor)
+
+
+@pytest.mark.parametrize('Model, Tokenizer, model_path',[
+    (AutoModelForCausalLM, AutoTokenizer, os.environ.get('FALCON_7B_ORIGIN_PATH'))
+])
```

Comment:
Can we use this to combine llama and falcon?
is the only diff in layer_name？

Reply:
yes I am prepared to combine them in next commit.